storage/range.go

```diff
@@ -271,30 +279,89 @@ func (r *Range) IsFirstRange() bool {
 	return bytes.Equal(r.Desc().StartKey, engine.KeyMin)
 }
 
-// IsLeader returns true if this range replica is the raft leader.
-// TODO(spencer): this is always true for now.
-func (r *Range) IsLeader() bool {
-	return true
+// getLease returns the current leader lease.
+func (r *Range) getLease() *proto.Lease {
+	return (*proto.Lease)(atomic.LoadPointer(&r.lease))
+}
+
+// newNotLeaderError returns a NotLeaderError intialized with the
+// replica for the last known holder of the leader lease.
+func (r *Range) newNotLeaderError() error {
+	err := &proto.NotLeaderError{}
+	if l := r.getLease(); l != nil {
+		_, storeID := DecodeRaftNodeID(multiraft.NodeID(l.RaftNodeID))
+		err.Leader = r.Desc().FindReplica(storeID)
+	}
+	return err
 }
 
-// canServiceCmd returns an error in the event that the range replica
-// cannot service the command as specified. This is of the case in
-// the event that the replica is not the leader.
-func (r *Range) canServiceCmd(method string, args proto.Request) error {
-	if !r.IsLeader() {
-		if !proto.IsReadOnly(method) || args.Header().ReadConsistency == proto.CONSISTENT {
-			// TODO(spencer): when we happen to know the leader, fill it in here via replica.
-			return &proto.NotLeaderError{}
+// HasLeaderLease returns whether the leader lease is held by this
+// replica and the term in which it was held. It's possible that the
+// replica is not in fact the leader, but holding the lease means that
+// no other replica may be the leader either. Leases may not overlap,
+// though a gap between successive lease holders is expected.
+func (r *Range) HasLeaderLease() (bool, uint64) {
+	if l := r.getLease(); l != nil {
+		if l.RaftNodeID == uint64(r.rm.RaftNodeID()) &&
+			r.rm.Clock().PhysicalNow() < l.Expiration {
+			return true, l.Term
 		}
+		return false, l.Term
 	}
-	if proto.IsReadOnly(method) {
-		if args.Header().ReadConsistency == proto.CONSENSUS {
-			return util.Errorf("consensus reads not implemented")
-		} else if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Txn != nil {
-			return util.Errorf("cannot allow inconsistent reads within a transaction")
+	return false, 0
+}
+
+// AnotherHasLeaderLease returns whether a replica other than this one
+// is known to be the holder of the lease. We use a grace period to
+// drown out clock offsets and provide hysteresis.
+func (r *Range) AnotherHasLeaderLease() bool {
```

Comment:
Why do we need this function? I thought that when updating `r.lease`, we'll substitute the local timestamp + duration into the expiration, so `lease.Expiration` will, in absolute time, already extend past the period in which the lease is held at the grantee and so we shouldn't have to worry about clock skew or add extra margins here.